projects/tsq-gallery/src/lib/components/tsq-gallery-viewer/tsq-gallery-viewer.component.ts

```diff
@@ -376,118 +259,150 @@ export class TSqGalleryViewerComponent {
-    aTag.click();
-    aTag.remove();
+  private adjustHorizontalPositionOnZoom() {
+    switch (this.imageRotation % 360) {
+      case 0:
+      case 180:
+      case -180:
+      case -360: {
+        this.positionLeft = this.imageZoom > 1.0 ? this.positionLeft / 2 : 0;
+        break;
+      }
+      default: {
+        this.positionTop = this.imageZoom > 1.0 ? this.positionTop / 2 : 0;
+        break;
+      }
+    }
   }
 
-  goFoward() {
-    if (this.canGoFoward) {
-      this.resetPosition();
-      this.selectedFileIndex++;
-      this.pdfLoading = false;
+  private moveImageTo(direction: SupportedScrollDirection, distance: number) {
+    const directionMod = direction === SupportedScrollDirection.UP || direction === SupportedScrollDirection.LEFT ? 1 : -1;
+    const nextDistance = this.addZoomModifier(distance);
+    const imageRotationMod = direction === SupportedScrollDirection.LEFT || direction === SupportedScrollDirection.RIGHT ? 90 : 0;
 
-      this.navigationDelay = true;
-      setTimeout(() => this.navigationDelay = false, 250);
+    switch ((this.imageRotation + imageRotationMod) % 360) {
+      case 0:
+      case -360: {
+        this.positionTop += directionMod * nextDistance;
+        break;
+      }
+      case 90:
+      case -270: {
+        this.positionLeft += directionMod * nextDistance;
+        break;
+      }
+      case 180:
+      case -180: {
+        this.positionTop += -directionMod * nextDistance;
+        break;
+      }
+      case 270:
+      case -90: {
+        this.positionLeft += -directionMod * nextDistance;
+        break;
+      }
     }
   }
 
-  goBack() {
-    if (this.canGoBack) {
-      this.resetPosition();
-      this.selectedFileIndex--;
-      this.pdfLoading = false;
+  private resetImagePosition(resetRotation: boolean = true) {
+    this.setImageZoom(1, true);
+    this.initialX = this.initialY = this.initialLeft = this.initialTop = this.positionLeft = this.positionTop = 0;
 
-      this.navigationDelay = true;
-      setTimeout(() => this.navigationDelay = false, 250);
+    if (resetRotation) {
+      this.imageRotation = 0;
     }
   }
 
-  closeBackdrop($event: TouchEvent | MouseEvent) {
-    $event.stopPropagation();
+  private setCurrentFile(index: number) {
+    this.resetImagePosition();
 
-    if (this.backdropClickClose && !this.isMoving && !this.showLoading) {
-      close();
-    }
+    this.pdfLoading = false;
+    this.selectedFileIndex = index || 0;
+    this.selectedFileToDisplay = !!this.galleryFiles && this.galleryFiles[index || 0];
+
+    this.canGoBack = this.selectedFileIndex > 0;
+    this.canGoForward = this.selectedFileIndex  < this.galleryFiles.length - 1;
+
+    this.topViewerContext = {
+      file: this.selectedFileToDisplay,
+      index: this.selectedFileIndex,
+      fns: {
+        close: this.close,
+      },
+    };
+
+    this.bottomViewerContext = {
+      file: this.selectedFileToDisplay,
+      index: this.selectedFileIndex,
+    };
   }
 }
```

Comment:
This file is huge. We need to think a way of extracting it to other files, components or find other alternative to keep this cleaner. I think extracting the pure functions with the same scope to some utility files would do some of the work, but maybe other approaches are needed.

Reply:
I was thinking on doing something like that, but I came to the conclusion that I would end up just moving literally everything to that class which would result in the same problem there

Reply:
I will isolate the image positioning behaviors into a model file